storage/file_persister_test.go

```diff
@@ -74,3 +79,174 @@ func TestLocalFilePersister(t *testing.T) {
+			wantPresignedHeaders: map[string]string{
+				"Authorization": "token asd123",
+				"Run_id":        "123456",
+			},
+			getPresignedURLResponse: http.StatusInternalServerError,
+			wantError:               "get presigned url: status code out of range: status code is 500",
+		},
+		{
+			name: "upload_rate_limited",
+			path: "some/path/file.png",
+			data: "here's some data",
+			wantPresignedURLBody: `{
+					"service":"aws_s3",
+					"files":[{"name":"screenshots/some/path/file.png"}]
+				}`,
+			wantPresignedHeaders: map[string]string{
+				"Authorization": "token asd123",
+				"Run_id":        "123456",
+			},
+			uploadResponse:          http.StatusTooManyRequests,
+			getPresignedURLResponse: http.StatusOK,
+			wantError:               "upload status code out of range: request is being rate limited",
+		},
+		{
+			name: "upload_fails",
+			path: "some/path/file.png",
+			data: "here's some data",
+			wantPresignedURLBody: `{
+					"service":"aws_s3",
+					"files":[{"name":"screenshots/some/path/file.png"}]
+				}`,
+			wantPresignedHeaders: map[string]string{
+				"Authorization": "token asd123",
+				"Run_id":        "123456",
+			},
+			uploadResponse:          http.StatusInternalServerError,
+			getPresignedURLResponse: http.StatusOK,
+			wantError:               "upload status code out of range: status code is 500",
+		},
+	}
+
+	for _, tt := range tests {
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
+
+			// This starts the upload server where the files will be uploaded
+			// with the presigned url.
+			uts := httptest.NewServer(http.HandlerFunc(
+				func(w http.ResponseWriter, r *http.Request) {
+					defer func() {
+						require.NoError(t, r.Body.Close())
+					}()
+
+					bb, err := io.ReadAll(r.Body)
+					require.NoError(t, err)
+
+					// Ensures that the data is uploaded to the server and matches
+					// what was sent.
+					assert.Equal(t, tt.data, string(bb))
+
+					w.WriteHeader(tt.uploadResponse)
+				}))
+			defer uts.Close()
+
+			// This starts the presigned url server which will return the
+			// presigned url for the given file name.
+			pts := httptest.NewServer(http.HandlerFunc(
+				func(w http.ResponseWriter, r *http.Request) {
+					defer func() {
+						require.NoError(t, r.Body.Close())
+					}()
+
+					bb, err := io.ReadAll(r.Body)
+					require.NoError(t, err)
+
+					// Ensures that the body of the request matches the
+					// expected format.
+					assert.JSONEq(t, tt.wantPresignedURLBody, string(bb))
+
+					// Ensures that the headers are sent to the server from
+					// the browser module.
+					for k, v := range tt.wantPresignedHeaders {
+						assert.Equal(t, v, r.Header[k][0])
+					}
+
+					body := fmt.Sprintf(`{
+						"service": "aws_s3",
+						"urls": [{
+							"name": "%s",
+							"pre_signed_url": "%s"
+						}]
+						}`, basePath, uts.URL)
+
+					w.WriteHeader(tt.getPresignedURLResponse)
+					_, err = w.Write([]byte(body))
+					require.NoError(t, err)
```

Comment:
`fmt.Fprintf`?

Reply:
Resolved in 7dd12ad0451b5f4f917814ded786174d78a9837d